### PR TITLE
test(plugin):  add tests

### LINKED ISSUE #105: Unable to load gzipped version of import()ed assets
I am working on a project at @thumbtack that gzips assets that internally use the ES6 dynamic `import()`. 

However, I cannot find a configuration of webpack/this specific plugin that allows me to load the dynamically generated assets as their gzipped version. If I specify the webpack `output.filename` field as `[name].js.gz` (which feels a bit like a hack), then I have to specify `asset: '[path][query]'` in this plugin's config and - more importantly - the files are no longer properly minified by webpack.

Frankly I couldn't quite figure out if this issue belongs here or in [webpack proper](https://github.com/webpack/webpack/). It seems to live somewhere in the middle.

### Minimal repo

https://github.com/bawjensen/webpack-gzip-import-issue

### PR DESCRIPTION
**Pull request by evilebottnawi**
### Type

- [x] Test

### Issues

- Fixes #105

### SemVer

- [x] Patch

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -13,7 +13,7 @@
     "appveyor:test": "npm run test",
     "build": "cross-env NODE_ENV=production babel src -d dist --ignore 'src/**/*.test.js'",
     "clean": "del-cli dist",
-    "lint": "eslint --cache src test",
+    "lint": "eslint --cache src test --ignore-pattern fixtures",
```
Comment (michael-ciniawsky) [inline package.json]: Could this be declared in the `eslintrc.js` instead ?

File: src/index.js [hunk 2]
```diff
@@ -124,7 +124,7 @@ class CompressionPlugin {
             const sub = {
               file,
               path: parse.pathname,
-              query: parse.query || '',
+              query: parse.query ? `?${parse.query}` : '',
```
Comment (michael-ciniawsky) [inline src/index.js]: Why is this change needed ? Bugfix ?

File: test/CompressionPlugin.test.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
+import Plugin from '../src/index';
+import { cleanErrorStack, createCompiler, compile } from './helpers';
```
Comment (michael-ciniawsky) [inline test/CompressionPlugin.test.js]: Ideally we agree to use `@webpacl-contrib/test-utils` to keep the cpmmon helpers in one place in the future, this can be done later as e.g the `cleanErrorStack` helper is currently missing there

File: test/CompressionPlugin.test.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+      output: {
+        path: `${__dirname}/dist`,
+        filename: '[name].js?var=[hash]',
+        chunkFilename: '[id].[name].js?ver=[hash]',
+      },
+    });
+  });
+
+  it('should works (without options)', () => {
+    new Plugin().apply(compiler);
```
Comment (michael-ciniawsky) [inline test/CompressionPlugin.test.js]:
```js
const config = {
   plugins: [
      new Plugin(options)
   ]
}
```

File: package.json (hunk 1 above)
Comment (evilebottnawi) [reply package.json]: @michael-ciniawsky let's discuss about this in `webpack-defaults`

File: src/index.js (hunk 2 above)
Comment (evilebottnawi) [reply src/index.js]: Bug fix https://github.com/webpack-contrib/compression-webpack-plugin/issues/105

File: test/CompressionPlugin.test.js (hunk 4 above)
Comment (evilebottnawi) [reply test/CompressionPlugin.test.js]: @michael-ciniawsky no, it is impossible test some features like errors inside `stats`

File: test/CompressionPlugin.test.js (hunk 3 above)
Comment (evilebottnawi) [reply test/CompressionPlugin.test.js]: @michael-ciniawsky `@webpacl-contrib/test-utils` is good for testing loader, but many plugin require own environment, `test-utils` is vacuum implementation that does not fit in most cases

**Comment by evilebottnawi** [comment]
@michael-ciniawsky i disagree about structure of tests, when you writing tests you should keep test file in place where you have file code (`src/plugin/file.js` => `test/plugin/file.test.js`), create `options` folder is break this best practice.

File: package.json (hunk 1 above)
Comment (michael-ciniawsky) [reply package.json]: https://github.com/webpack-contrib/webpack-defaults/pull/130

File: src/index.js (hunk 2 above)
Comment (michael-ciniawsky) [reply src/index.js]: Always better to do this in a separate PR for `CHANGELOG` generation reasons and to keep things clear

Comment (evilebottnawi) [reply src/index.js]: @michael-ciniawsky it is break many tests and it is very inconvenient in this case, also it is not difficult write CHANGELOG manually in this cases

**Comment by michael-ciniawsky** [comment]
Since `options` aren't in separate files I'm not following, `options/` just groups them in one place and removes the need for redundant test naming `x-option.test.js`. Alternatively 

```
test
|– ...
|– Errors.test.js
|– defaults.test.js
|– option.test.js
|– ...
```

to keep it flat

**Comment by evilebottnawi** [comment]
@michael-ciniawsky 
- Keep all options `option.test.js` will create just a huge file with tests.

Don't see problems `x-options.test.js`:
- Easy single test `node_modules/bin/jest test/x-options.test.js` 
- Easy add/remove when you add/remove option
- Use best practice for test places - i.e. `src/plugin/file.js` -> `test/plugin/file.js` (it is recommendation not only for `js`, also for `php/ruby/go`)

**Comment by michael-ciniawsky** [comment]
> Keep all options option.test.js will create just a huge file with tests.

No I still meant in to keep them in a file per option of course 


> P.S.
Actually, I wanted today release compression-webpack-plugin@2 with updated deps, fixed some problems and some improvements. If we did not waste time on stupid things (like naming, newlines, places of files) and write more code, I would have done it.

Why does it need to be **today**? Where is it coming from that you often ping someone for review and put 'a knife on his/her's throat to please do it fast' ? 🙃 Please chill...

> If we did not waste time on stupid things (like naming, newlines, places of files) and write more code, I would have done it.

Do you think I personally like this shit ? I started to write `test-utils` to not having to discuss this stuff in the future, for e.g why aren't you using them here ? 🙃 I wrote them for now since contributions are currenty **zero**  there, which I can't possibly affect... But there shouldn't be anything opinionated atm since I tried to avoid it at all cost 

> Therefore, from the current moment we are discussing all these things in webpack-defaults (eslint/prettier/etc configuration to avoid typo in code and docs, approves how it should look README and structure of tests and etc). And gradually updates webpack-defaults on each repo in webpack-contrib.

The current `webpack-defaults` has to much **unnecessary** tooling, which caused me more problems than it solved and the same is true for the current `eslint` config. I never in my development career needed to write so many `/* eslint-disable */` comments... I fully agree that we should solve that per standardization to just focus on the relevant parts. The fact is that having this kind of workflow isn't currently true and you also do what you like in terms of testing atm, while claiming others shouldn't be so pendantic when reviewing your PR's... If you and I and potentially others always slightly do this or that differently then either this is accepted as a matter of fact or one has to make a note 

- [x] Refactor tests 

to at least try to keep things consistent from time to time

**Comment by evilebottnawi** [comment]
@michael-ciniawsky 
> Why does it need to be today? Where is it coming from that you often ping someone for review and put 'a knife on his/her's throat to please do it fast' ? upside_down_face Please chill...

Because time is what we have and it is expensive, each discussion costs expensive for me and other developers. I have regular work and can't spent a lot of time on minor things (naming, test structures and etc).

> why aren't you using them here

Because i will try use their in `terser/uglify/compression/etc` plugin and they do not have enough options for customization. 

I merge this as is, also here no `webpack-defaults` update and in this version no `test-utils`. I will migrate on this in next major release.
